src/main/scala/mesosphere/marathon/tasks/TaskQueue.scala

```diff
@@ -17,21 +18,31 @@ class TaskQueue {
 
   import mesosphere.marathon.tasks.TaskQueue._
 
+  private val log = Logger.getLogger(getClass)
   protected[marathon] val rateLimiter = new RateLimiter
 
-  // we used SynchronizedPriorityQueue before, but it has been deprecated
-  // because it is not safe to use
-  protected[tasks] var queue =
-    new PriorityBlockingQueue[QueuedTask](11, AppConstraintsOrdering.reverse)
+  protected[tasks] var apps = TrieMap.empty[(PathId, Timestamp), QueuedTask]
 
-  def list: Seq[QueuedTask] = queue.asScala.to[scala.collection.immutable.Seq]
+  def list: Seq[QueuedTask] = apps.values.to[Seq]
 
   def listApps: Seq[AppDefinition] = list.map(_.app)
 
-  def poll(): Option[QueuedTask] = Option(queue.poll())
-
-  def add(app: AppDefinition): Unit =
-    queue.add(QueuedTask(app, rateLimiter.getDelay(app)))
+  def poll(): Option[QueuedTask] =
+    apps.values.toSeq.sortWith {
+      case (a, b) =>
+        a.app.constraints.size > b.app.constraints.size
+    }.find {
+      case QueuedTask(_, count, _) => count.decrementAndGet() >= 0
+    }
+
+  def add(app: AppDefinition): Unit = add(app, 1)
+
+  def add(app: AppDefinition, count: Int): Unit = {
+    val queuedTask = apps.getOrElseUpdate(
+      (app.id, app.version),
+      QueuedTask(app, new AtomicInteger(0), rateLimiter.getDelay(app)))
```

Comment:
Oh, that's confusing. The RateLimiter takes an app, but only uses the id for the mapping. I think it should be versioned. That way we don't have to modify the delay in the queue. 



> On 06.01.2015, at 18:45, Connor Doyle <notifications@github.com> wrote:
> 
> In src/main/scala/mesosphere/marathon/tasks/TaskQueue.scala:
> 
> > -  def add(app: AppDefinition): Unit =
> > -    queue.add(QueuedTask(app, rateLimiter.getDelay(app)))
> > +  def poll(): Option[QueuedTask] =
> > +    apps.values.toSeq.sortWith {
> > +      case (a, b) =>
> > +        a.app.constraints.size > b.app.constraints.size
> > +    }.find {
> > +      case QueuedTask(_, count, _) => count.decrementAndGet() >= 0
> > +    }
> > +
> > +  def add(app: AppDefinition): Unit = add(app, 1)
> > +
> > +  def add(app: AppDefinition, count: Int): Unit = {
> > +    val queuedTask = apps.getOrElseUpdate(
> > +      (app.id, app.version),
> > +      QueuedTask(app, new AtomicInteger(0), rateLimiter.getDelay(app)))
> Previously, a new deadline would be added to each QueuedTask. Now, adding a number of tasks to the queue will only update the count if there are items in the queue for that app. So, a likely scenario is that we enquire a large number of tasks for a badly specified app, we launch several, they fail (increasing the delay in the rate limiter), and the next tasks added will have the old deadline instead of one derived from the new delay. We should replace the deadline with a new one, or make a new QueuedTask for the new batch of tasks with a new deadline. Does that make sense?
> 
> —
> Reply to this email directly or view it on GitHub.
>